source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,2515 @@
+
+  This field is mutually exclusive with
+  `expectedError <operation_expectedError_>`_.
+
+  This is primarily used for creating a `changeStream`_ entity from the result
+  of a `client_createChangeStream`_, `database_createChangeStream`_, or
+  `collection_createChangeStream`_ operation.
+
+
+expectedError
+~~~~~~~~~~~~~
+
+One or more assertions for an error/exception, which is expected to be raised by
+an executed operation. At least one key is required in this document.
+
+The structure of this document is as follows:
+
+- ``isError``: Optional boolean. If true, the test runner MUST assert that an
+  error was raised. This is primarily used when no other error assertions apply
+  but the test still needs to assert an expected error. Test files MUST NOT
+  specify false, as `expectedError`_ is only applicable when an operation is
+  expected to raise an error.
+
+- ``isClientError``: Optional boolean. If true, the test runner MUST assert that
+  the error originates from the client (i.e. it is not derived from a server
+  response). If false, the test runner MUST assert that the error does not
+  originate from the client.
+
+  Client errors include, but are not limited to: parameter validation errors
+  before a command is sent to the server; network errors.
+
+- ``errorContains``: Optional string. A substring of the expected error message
+  (e.g. "errmsg" field in a server error document). The test runner MUST assert
+  that the error message contains this string using a case-insensitive match.
+
+  See `bulkWrite`_ for special considerations for BulkWriteExceptions.
+
+- ``errorCode``: Optional integer. The expected "code" field in the
+  server-generated error response. The test runner MUST assert that the error
+  includes a server-generated response whose "code" field equals this value.
+  In the interest of readability, YAML files SHOULD use a comment to note the
+  corresponding code name (e.g. ``errorCode: 26 # NamespaceNotFound``).
+
+  Server error codes are defined in
+  `error_codes.yml <https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml>`__.
+
+  Test files SHOULD NOT assert error codes for client errors, as specifications
+  do not define standardized codes for client errors.
+
+- ``errorCodeName``: Optional string. The expected "codeName" field in the
+  server-generated error response. The test runner MUST assert that the error
+  includes a server-generated response whose "codeName" field equals this value
+  using a case-insensitive comparison.
+
+  See `bulkWrite`_ for special considerations for BulkWriteExceptions.
+
+  Server error codes are defined in
+  `error_codes.yml <https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml>`__.
+
+  Test files SHOULD NOT assert error codes for client errors, as specifications
+  do not define standardized codes for client errors.
+
+- ``errorLabelsContain``: Optional array of strings. A list of error label
```

Comment:
Using matchers for errors rather then specialized keys.

Currently we've got pretty specific error matching properties `errorContains`,
`errorLabelsContain`, `errorLabelsOmit`. I can imagine more being added as time
goes on, I was wondering the same convention we use to match results or a
document, if we can match parts of an error the same way.

```js
{
  error: {
    $$exists: true,
    message: {
      $$contains: 'hello',
    }
    code: {
      $$equals: 91,
    }
    labels: {
      $$contains: ['RetryableWrite'],
      $$omits: ['TimeoutError'],
    }
  }
}
```

The idea here being that an implementation can create a singular `match`
function as described in the doc, and pass parts of an error through. This could
also possible fix our issue with just having and error `$$exists` by allowing
top level matchers.

Reply:
Great explanation. I may have mention this in a previous response, but my way of thinking here is more using driver-specific-mappings to map error objects, to their designated property or type. And as for the matchers it's to have a standardized way of expecting results. I do think driver-specific-mappings is more overhead, as well as error-specific matchers and understand your rational here.